src/stores/right-panel/RightPanelStore.ts

```diff
@@ -51,20 +51,21 @@ export default class RightPanelStore extends ReadyWatchingStore {
     } = {};
     private viewedRoomId: Optional<string>;
 
-    private constructor() {
+    // Public for tests
+    public constructor() {
         super(defaultDispatcher);
     }
 
-    protected async onReady(): Promise<any> {
+    protected onReady = async (): Promise<any> => {
```

Comment:
why do these need to be bound functions?

Reply:
Because otherwise `onReady` gets called from *the constructor* of the parent class, which apparently interacts weirdly with the overriding and binding of `onReady`, causing a type error that `this.onVerificationRequestUpdate` is undefined.

Reply:
err, that's extremely surprising. It being called from the parent class shouldn't cause `this` to change scope in a weird way like this. 

it also wasn't a bound function before, which has me quite curious if it ever worked. 

Do you have a stacktrace from when it failed?

Reply:
Yeah, it's confused me too, but if it's not bound, then 10 out of the 19 RightPanelStore tests fail in the `beforeEach` stage with:

```
TypeError [ERR_INVALID_ARG_TYPE]: The "listener" argument must be of type function. Received undefined

  58 |     protected async onReady(): Promise<any> {
  59 |         this.viewedRoomId = RoomViewStore.instance.getRoomId();
> 60 |         this.matrixClient.on(CryptoEvent.VerificationRequest, this.onVerificationRequestUpdate);
     |                           ^
  61 |         this.loadCacheFromSettings();
  62 |         this.emitAndUpdateSettings();
  63 |     };

  at RightPanelStore.onReady (src/stores/right-panel/RightPanelStore.ts:60:27)
  at new ReadyWatchingStore (src/stores/ReadyWatchingStore.ts:40:18)
  at new RightPanelStore (src/stores/right-panel/RightPanelStore.ts:55:9)
  at Object.<anonymous> (test/stores/right-panel/RightPanelStore-test.ts:45:17)
```

and I have no idea what to make of that. Same thing for the RightPanel test suite, in which both tests fail `at spinUpStores (test/components/structures/RightPanel-test.tsx:76:25)`.

Reply:
It hurts my brain to think about how that could possibly happen, but binding the methods seems to work and be benign, so :woman_shrugging:

Reply:
The bound function seems to change the call path a bit. Walking through it with a debugger shows that the construction of a `RightPanelStore` sends it through the `super`'s `onReady`, which is the wrong `onReady`. It's only the manual call in the test which hits the actual `onReady` function defined by `RightPanelStore`. Switching it back to an unbound function means the `RightPanelStore`'s override is in fact called, though with a weird `this` reference which appears to be a different `RightPanelStore` entirely... 

/me pokes it some more.

Reply:
Right, so it's because of how much of a nightmare our store layout is. 

TLDR: This fixes it:

```diff
diff --git a/src/stores/right-panel/RightPanelStore.ts b/src/stores/right-panel/RightPanelStore.ts
index 2cfafb26ac2..b3e46e7ff71 100644
--- a/src/stores/right-panel/RightPanelStore.ts
+++ b/src/stores/right-panel/RightPanelStore.ts
@@ -53,18 +53,28 @@ export default class RightPanelStore extends ReadyWatchingStore {

     private constructor() {
         super(defaultDispatcher);
+        this.reset();
     }

-    protected onReady = async (): Promise<any> => {
+    protected async onReady(): Promise<any> {
         this.viewedRoomId = RoomViewStore.instance.getRoomId();
         this.matrixClient.on(CryptoEvent.VerificationRequest, this.onVerificationRequestUpdate);
         this.loadCacheFromSettings();
         this.emitAndUpdateSettings();
-    };
+    }

-    protected onNotReady = async (): Promise<any> => {
+    protected async onNotReady(): Promise<any> {
         this.matrixClient.off(CryptoEvent.VerificationRequest, this.onVerificationRequestUpdate);
-    };
+    }
+
+    /**
+     * Resets the store. Intended for test usage only.
+     */
+    public reset() {
+        this.global = null;
+        this.byRoom = {};
+        this.viewedRoomId = null;
+    }

     protected onDispatcherAction(payload: ActionPayload) {
         if (payload.action !== Action.ActiveRoomChanged) return;
diff --git a/test/components/structures/RightPanel-test.tsx b/test/components/structures/RightPanel-test.tsx
index 9653fa6cd12..17e952e7fbd 100644
--- a/test/components/structures/RightPanel-test.tsx
+++ b/test/components/structures/RightPanel-test.tsx
@@ -70,9 +70,8 @@ describe("RightPanel", () => {
         // This is private but it's the only way to selectively enable stores
         await WidgetLayoutStore.instance.onReady();

-        // @ts-ignore
         // Make sure we start with a clean store
-        RightPanelStore.internalInstance = null;
+        RightPanelStore.instance.reset();
         RightPanelStore.instance.useUnitTestClient(cli);
         // @ts-ignore
         await RightPanelStore.instance.onReady();
```

### BUT HOW!? 
(not-so-TLDR)

The way the stores are structured means they are instantiated on import, generating an instance reference which then gets bound to the `window` object, just as a true singleton should. By resetting the `internalInstance` by hand then calling `useUnitTestClient()` we end up generating a new instance which detects the client on its own and uses that reference (because `MatrixClientPeg.get()` is mocked). Because it detects an instance, and because the singletons are now confused (there's two `RightPanelStore`s floating around: one instantiated and one attached to the `window`), `this` within `onReady` actually has a `ReadyWatchingStore` prototype, hence the undefined function call. 

The confusing part was trying to figure out why this didn't affect the `WidgetLayoutStore` given it's able to figure out which of the `onReady` functions to call, and with the correct context. It also has a `window`-stored reference created at the time of import, however its `internalInstance` wasn't being reset - if you add in a similar `internalInstance = null` for the `WidgetLayoutStore` then you'll mimic the bug the `RightPanelStore` was having.

To work around the issue, we just create a `reset()` function which obliterates state. For safety and to encourage developers to set sane defaults, we call that reset function in the constructor so that if we forget a class member during a refactoring then it'll have a high chance of breaking something (either at runtime or in tests) - eg if we forgot to reset `byRoom` then we might find conflicts between tests and therefore some angry error messages.

Reply:
Hmm, okay, thanks for investigating (and reminding me that arrow functions do not actually provide the desired override semantics). I still really can't follow this bit of the logic:

> Because it detects an instance, and because the singletons are now confused (there's two RightPanelStores floating around: one instantiated and one attached to the window), this within onReady actually has a ReadyWatchingStore prototype

i.e. how having two instances of a class would possibly impact the prototypes of those instances, even if they're semantically meant to be singletons. However I definitely see how your suggestion would be overall a more sane thing to do.

Reply:
that much I haven't figured out, but I think it's effectively call order and mocks getting confused, somehow.